src/main/java/io/debezium/connector/vitess/VitessConnector.java

```diff
@@ -41,10 +46,49 @@ public Class<? extends Task> taskClass() {
 
     @Override
     public List<Map<String, String>> taskConfigs(int maxTasks) {
-        if (maxTasks > 1) {
-            throw new IllegalArgumentException("Only a single connector task may be started");
+        LOGGER.info("Calculating taskConfigs for {} tasks", maxTasks);
+        List<String> shards = null;
+        String storagePerTask = properties.get(VitessConnectorConfig.OFFSET_STORAGE_PER_TASK.name());
+        if (storagePerTask != null && storagePerTask.equalsIgnoreCase("true")) {
+            VitessConnectorConfig connectorConfig = new VitessConnectorConfig(Configuration.from(properties));
+            shards = getVitessShards(connectorConfig);
+        }
+        return taskConfigs(maxTasks, shards);
+    }
+
+    public List<Map<String, String>> taskConfigs(int maxTasks, List<String> shards) {
+        LOGGER.info("Calculating taskConfigs for {} tasks and shards: {}", maxTasks, shards);
+        String storagePerTask = properties.get(VitessConnectorConfig.OFFSET_STORAGE_PER_TASK.name());
+        if (storagePerTask != null && storagePerTask.equalsIgnoreCase("true")) {
+            int tasks = Math.min(maxTasks, shards.size());
+            LOGGER.info("There are {} vitess shards for maxTasks: {}, we will use {} tasks",
+                    shards.size(), maxTasks, tasks);
+            shards.sort(Comparator.naturalOrder());
+            Map<Integer, List<String>> shardsPerTask = new HashMap<>();
+            int taskId = 0;
+            for (String shard : shards) {
+                List<String> taskShards = shardsPerTask.computeIfAbsent(taskId, k -> new ArrayList<>());
+                taskShards.add(shard);
+                taskId = (taskId + 1) % tasks;
+            }
+            LOGGER.info("Shards task distribution: {}", shardsPerTask);
+            List<Map<String, String>> allTaskProps = new ArrayList<>();
+            for (int tid : shardsPerTask.keySet()) {
+                List<String> taskShards = shardsPerTask.get(tid);
+                Map<String, String> taskProps = new HashMap<>(properties);
+                taskProps.put(VitessConnectorConfig.VITESS_TASK_KEY_CONFIG, String.format("task%d_%d", tid, tasks));
```

Comment:
What's purpose having a task key property with task count in the key?
I remember in Kafka Connect, connector task id are integer only right?
Is it for discriminating the shard setup under the certain task count only since once chosen it cannot be changed unless we do a migration.

Reply:
The reason I've put in both taskId and total_num_tasks was trying to handle the offset migration during task number change (or leave enough clues for other people to write a tool to handle the auto migration).

Kafka Connect's OffsetStorageReader class only has API to read a specific key, it doesn't have the API to list all the keys in the storage.  During a task number change (e.g. from tasks.max = 2 to tasks.max = 4), I need to figure out all the previous partition keys corresponding to tasks.max=2 generation, I was intending to have the user specify prev.tasks.max=2 in the config, so I will know all the keys for previous generation are: task0_2, task1_2.  This way I can read all the old vgtids for those key and reshuffle them for tasks.max=4.   And for this generation, I will generate task0_4, task1_4, task2_4, task2_4.  So the keys from two generation don't overlap.   If I don't have have the extra _4 in the key name, the partition keys between two generations will collide on task0 and task1.  We might enter into a situation some task didn't get the time to save new VGtid for task0 before it was killed accidentally.  On restart, it will retrieve the VGtid for task0 from offset storage which belong to the old generation.

But I am currently not handling offset auto migration nor do I use total_number_tasks in the task key, so I am open to either ways.  But I feel having total_number_tasks in the partition key name clearly shows which generation this partitioning belongs, it might be good for debugging as well.